File: lab4/Figures/Domain/Compound/Compound.cpp
#include "Compound.hpp"
#include "../../../../Utils/StringUtils.hpp"

#include <iostream>

double Compound::GetDensity() const
{
    return GetFromAll([](body body) { return body->GetDensity(); });
}

double Compound::GetVolume() const
{
    return GetFromAll([](body body) { return body->GetVolume(); });
}

double Compound::GetMass() const
{
    return GetFromAll([](body body) { return body->GetMass(); });
}

void Compound::AddBody(std::shared_ptr<Body> body)
{
    if (typeid(Compound).hash_code() == typeid(*body).hash_code())
    {
        if (CheckForPresence(body))
        {
            return;
        }

        std::cout << "I don`t know what to say" << std::endl;
        Compound *compound = dynamic_cast<Compound*>(body.get());
        compound->parentBody = std::shared_ptr<Compound>(this);
    }

    bodies.push_back(body);
    std::cout << "Temp crash" << std::endl;
}

Compound::operator std::string() const
{
    return StringUtils::StringConcatenator() << GetDensity() <<  ' ' << GetVolume() << ' ' << GetMass();
}

double Compound::GetFromAll(std::function<double(std::shared_ptr<Body>)> getter) const
{
    double sum;

    for (size_t i = 0; i < bodies.size(); i++)
    {
        sum += getter(bodies[i]);
    }

    return sum;
}

bool Compound::CheckForPresence(body body)
{
    if (parentBody == nullptr)
    {
        return false;
    }

    if (parentBody.get() == body.get())
    {
        return true;
    }

    return parentBody->CheckForPresence(body);
}

File: lab4/Bodies/lib/Figures/Compound/Compound.dart
import 'package:BodiesLab/Figures/Compound/SelfInsertException.dart';

import '../Body/Body.dart';

class Compound implements Body
{
  final List<Body> _bodies = [];
  Compound _parentBody = null;

  double _GetFromAll(double Function(Body body) getter)
  {
    double result = 0;

    for (Body body in _bodies)
    {
      result += getter(body);
    }

    return result;
  }

  bool _CheckForPresence(Body body)
  {
    if (body == this)
    {
      return true;
    }

    if (_parentBody == null)
    {
      return false;
    }

    if (_parentBody == body)
    {
      return true;
    }

    return _parentBody._CheckForPresence(body);
  }

  double GetDensity()
  {
    return _GetFromAll((Body body) => body.GetDensity());
  }

  double GetVolume()
  {
    return _GetFromAll((Body body) => body.GetVolume());
  }

  double GetMass()
  {
    return _GetFromAll((Body body) => body.GetMass());
  }

  String toString() => 'Compound: \n' + _bodies.map((Body e) => e.toString()).join('\n');

  void AddBody(Body body)
  {
    if (body is Compound)
    {
      if (_CheckForPresence(body))
      {
        throw SelfInsertException();
      }

      body._parentBody = this;
    }

    _bodies.add(body);
  }
}
